File: pset6/readability/readability.py
from cs50 import get_string


def main():
    # get_string gets string input from user
    string = get_string("Text: ")

    # calculate number of letters and words in string
    letters = sum(char.isalpha() for char in string)
    spaces = sum(char.isspace() for char in string)
    words = spaces + 1

    # calculate number of sentend in string
    sentences = 0
    for char in string:
        if char in [".", "?", "!"]:
            sentences += 1

    # calculate Coleman-Liau index
    index = get_index(letters, words, sentences)

    # print reading grade
    print_index(index)


def get_index(letters, words, sentences):
    # Reading level is from Coleman-Liau index = 0.0588 * L - 0.296 * S - 15.8
    L = (letters / words) * 100
    S = (sentences / words) * 100
    index = round(0.0588 * L - 0.296 * S - 15.8)
    return index


def print_index(index):
    if index < 1:
        print("Before Grade 1")
    elif index > 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


if __name__ == "__main__":
    main()

File: pset2/Readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <math.h>

// declare functions at the top that will be used later in main
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
int get_index(int letters, int words, int sentences);

int main(void)
{
    // prompt user for text and store in a string variable called text
    string text = get_string("Text: ");

    // use custom-built functions to obtain number of letters, words and sentences
    int letters = count_letters(text);
    int words = count_words(text) + 1;
    int sentences = count_sentences(text);

    // use custom-built function to obtain reading grade and print grade
    int index = get_index(letters, words, sentences);
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

// function to count number of letters in text
int count_letters(string text)
{
    int counter = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (islower(text[i]) || isupper(text[i]))
        {
            counter++;
        }
    }
    return counter;
}

// function to count number of words in text
int count_words(string text)
{
    int counter = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (isspace(text[i]))
        {
            counter++;
        }
    }
    return counter;
}

// function to count number of sentences in text
int count_sentences(string text)
{
    int counter = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (text[i] == '.' || text[i] == '?' || text[i] == '!')
        {
            counter++;
        }
    }
    return counter;
}

// function to get reading level of text
// Coleman-Liau index = 0.0588 * L - 0.296 * S - 15.8
int get_index(int letters, int words, int sentences)
{
    float L = (float) letters / (float) words * 100;
    float S = (float) sentences / (float) words * 100;
    float index = round(0.0588 * L - 0.296 * S - 15.8);
    return index;
}